master/internal/db/postgres_tasks.go

```diff
@@ -437,6 +438,28 @@ WHERE task_id = $1
 	return count, nil
 }
 
+// DeleteExpiredTaskLogs deletes task logs older than the log_retention time where defined and
+// non-negative. If log_retention is not defined the defaultLogRetention value is used instead.
+func (db *PgDB) DeleteExpiredTaskLogs(defaultLogRetention int16) (int64, error) {
+	r, err := db.sql.Exec(fmt.Sprintf(`
+WITH log_retention_tasks AS (
+	SELECT task_id, end_time, COALESCE(log_retention, %d) AS current_retention FROM tasks
+	WHERE task_id IN (SELECT DISTINCT task_id FROM task_logs)
+		AND end_time IS NOT NULL
+)
+DELETE FROM task_logs
+WHERE task_id IN (
+	SELECT task_id FROM log_retention_tasks
+	WHERE current_retention >= 0
```

Comment:
as a case for NULL as "don't delete", i'd rather a null explode here than someone mess up and delete all the logs if they change this query. or, at least this query needs tests.